src/commands/signzone.rs

```diff
@@ -0,0 +1,285 @@
+                ))
+            })?;
+
+            let old_ext = key_path.extension().unwrap().to_str().unwrap();
+            let new_ext = format!("{}.key", old_ext);
+            let public_key_path = key_path.with_extension(new_ext).display().to_string();
+            let public_data = std::fs::read_to_string(&public_key_path).map_err(|err| {
+                Error::from(format!(
+                    "Unable to load public key from file '{}': {}",
+                    private_key_path, err
+                ))
+            })?;
+
+            let secret_key =
+                domain::sign::SecretKeyBytes::parse_from_bind(&private_data).map_err(|err| {
+                    Error::from(format!(
+                        "Unable to parse BIND formatted private key file '{}': {}",
+                        private_key_path, err
+                    ))
+                })?;
+
+            let public_key_info: domain::validate::Key<Bytes> =
+                domain::validate::Key::parse_from_bind(&public_data).map_err(|err| {
+                    Error::from(format!(
+                        "Unable to parse BIND formatted public key file '{}': {}",
+                        private_key_path, err
+                    ))
+                })?;
+
+            let key_pair = domain::sign::common::KeyPair::from_bytes(
+                &secret_key,
+                public_key_info.raw_public_key(),
+            )
+            .map_err(|err| {
+                Error::from(format!(
+                    "Unable to import private key from file '{}': {}",
+                    private_key_path, err
+                ))
+            })?;
+
+            let signing_key = domain::sign::SigningKey::new(
+                public_key_info.owner().to_owned(),
+                public_key_info.flags(),
+                key_pair,
+            );
+
+            keys.push(signing_key);
+        }
+
+        //---
+
+        let (apex, ttl) = Self::find_apex(&records).unwrap();
+
+        let opt_out = if self.hash_only {
+            Nsec3OptOut::OptOut
+        } else if self.nsec3_opt_out_flags_only {
+            Nsec3OptOut::OptOutFlagsOnly
+        } else {
+            Nsec3OptOut::NoOptOut
+        };
+
+        let hashes = if self.use_nsec3 {
+            let params = Nsec3param::new(self.algorithm, 0, self.iterations, self.salt.clone());
+            let Nsec3Records {
+                recs,
+                param,
+                hashes,
+            } = records
+                .nsec3s::<_, BytesMut>(&apex, ttl, params, opt_out, self.diagnostic_comments)
+                .unwrap();
+            records.extend(recs.into_iter().map(Record::from_record));
+            records.insert(Record::from_record(param)).unwrap();
+            hashes
+        } else {
+            let nsecs = records.nsecs::<Bytes>(&apex, ttl);
+            records.extend(nsecs.into_iter().map(Record::from_record));
+            None
+        };
+
+        if !self.hash_only {
```

Comment:
It isn't immediately clear what "not only hashing" means, could you add a comment saying that this block will actually insert signatures for the records?

Reply:
I'll introduce some sort of self-describing mode enum and use that instead.